FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/FreeRTOS_DNS.c

```diff
@@ -720,45 +1070,65 @@ TickType_t uxWriteTimeOut_ticks = ipconfigDNS_SEND_BLOCK_TIME_TICKS;
 			/* Get a buffer.  This uses a maximum delay, but the delay will be
 			capped to ipconfigUDP_MAX_SEND_BLOCK_TIME_TICKS so the return value
 			still needs to be tested. */
+			FreeRTOS_printf( ( "DNS lookup: \"%s\"\n", pcHostName ) );
 
 			uxHeaderBytes = ipSIZE_OF_ETH_HEADER + ipSIZE_OF_IPv4_HEADER + ipSIZE_OF_UDP_HEADER;
 
 			pxNetworkBuffer = pxGetNetworkBufferWithDescriptor( uxHeaderBytes + uxExpectedPayloadLength, 0UL );
 			if( pxNetworkBuffer != NULL )
 			{
 				pucUDPPayloadBuffer = &( pxNetworkBuffer->pucEthernetBuffer[ uxHeaderBytes ] );
+				/* Later when translating form UDP payload to a Network Buffer,
+				it is important to know whether this is an IPv4 packet. */
+				pucUDPPayloadBuffer[ - ( ( int ) ipUDP_PAYLOAD_IP_TYPE_OFFSET ) ] = ( uint8_t ) ipTYPE_IPv4;
```

Comment:
Why is there a '-' in the array index? Can there be a negative value in the array index? Or am I missing something?

Reply:
The minus is intentional and well-tested. Do you really want to know how it works?

In a IPv4 library, it is easy to jump from a UDP payload buffer to the beginning of a packet:
the distance is 14 + 20 + 8 = 42 bytes.

In a library that also has IPv6 packets, the distance can also be 14 + 40 + 8 = 62 bytes.

The frame type is written at the 12th byte of a packet, in the Ethernet Header:

	#define ipIPv4_FRAME_TYPE	0x0800
	#define ipIPv6_FRAME_TYPE	0x86DD

But we must know the length of the packet to find the frame type!  If the packet is IPv4, the IP/UDP headers are much shorter than 62 bytes.  you would look at a value far before pucEthernetBuffer, which is random.

I decided to look at the IP-header.  Both IPv4 and IPv6 headers start with a IP-type indication:

For IPv4, `ucVersionHeaderLength` equals 0x45 ( IP version 4, length 5 x 4 = 20 bytes ).
For IPv6, `ucVersionTrafficClass` equals 0x60 ( IP version 6, length is fixed at 40 ).

Now to find the IP-type, we will go back **40 + 8 = 48 bytes** from the UDP payload buffer.
If the packet is IPv6, we look at the IPv6 type field.
If the packet is IPv4, we land at 6 bytes before the start of `pucEthernetBuffer`, which is reserved for the IP-type.:

~~~c
    #define ipUDP_PAYLOAD_IP_TYPE_OFFSET    48
    pucUDPPayloadBuffer[ - ( ( int ) ipUDP_PAYLOAD_IP_TYPE_OFFSET ) ] = ( uint8_t ) ipTYPE_IPv4;
~~~

Luckily, the word at that offset is not yet used.  IPv6 packets will get that field set with the value `0x60`, IPv4 packets with 0x45.

If you've got lost by now, I understand.